### ISSUE TITLE
ENH/TST/MAINT: fft: follow-ups for array API support

### ISSUE THREAD

**Issue by lucascolley** [scipy/scipy#19257]
After gh-19005, here are some follow-ups:
- [ ] Update `doc/source/dev/api-dev/array_api.rst` to reflect support in `fft`.
- [ ] `arg_err_msg` should be given a better name (something like `xp_unsupported_param` but maybe more concise?) and moved to `_lib._array_api.py` for use in other submodules.
- [ ] Look at all of the non-standard functions (like `fft2`) and rewrite them in terms of standard functions where possible to enable GPU support. This includes decomposing `hfftn` into `fftn` and `hfft` https://github.com/scipy/scipy/pull/19005#discussion_r1326295380.
- [ ] Write an array-agnostic pad https://github.com/scipy/scipy/pull/19005#discussion_r1319201782.
- [ ] Clean up dtype conversions in tests, see https://github.com/scipy/scipy/pull/19251.
- [ ] The device keyword of `fftfreq` and `rfftfreq` is not tested since it only works with PyTorch currently. When NumPy supports it, we should make appropriate changes to `fftfreq` and `rfftfreq` and start testing with it.
- [ ] Once the libraries we are testing with are all compatible, the uses of `xp_test` can be changed to just `xp`.
- [ ] Test the `plan` parameter of the basic functions if it comes into use.

I will be looking at the first three points in the next few days. Anyone is welcome to take the rest on!

**Comment by ev-br** [comment]
All boxes are checked save `pad`, which is covered by https://data-apis.org/array-api-extra/generated/array_api_extra.pad.html
So it seems we can declare this done @lucascolley ?

**Comment by lucascolley** [comment]
wanna send a PR to use `xpx.pad` at https://github.com/scipy/scipy/blob/main/scipy/fft/tests/test_real_transforms.py#L40 ?

**Comment by ev-br** [comment]
Oh thanks, this is what the checkbox is about! Was not easy to find out from the issue description, I've to admit :-).
Now, the two uses of `np.pad` in tests are blocked on `mode="edge"`, so it's an enhancement request to `xpx.pad`, https://github.com/data-apis/array-api-extra/issues/811

I first thought it's the only thing that keeps the test cpu-only but apparently not only, as`dct`, `dst` and their ilk are all CPU-only. Why is this, what is the crucial difference between `fft` and `dct`?

-------------------------------

Going forward, it's be great to sharpen this issue so that it's clearer what is actually missing. WDYT about either closing this one + open a new one, or edit this one to 
- link to the `pad` usage + upstream xpx issue, so that it's easier actionable
- add a checkbox for adding CUDA support for discrete sine/cosine transforms (assuming it's doable?)
- potentially rename the issue to stress that it's a tracker for further enhancements for the Array API support in `scipy.fft`
?

**Comment by lucascolley** [comment]
> I first thought it's the only thing that keeps the test cpu-only but apparently not only, asdct, dst and their ilk are all CPU-only. Why is this, what is the crucial difference between fft and dct?

`dct` et al. are not in https://data-apis.org/array-api/latest/extensions/fourier_transform_functions.html.

> WDYT about either closing this one + open a new one

sounds fine if you want to do that

**Comment by ev-br** [comment]
Re new issue: having slept on it, I think the upstream issue, https://github.com/data-apis/array-api-extra/issues/811, is sufficient: once `xpx.pad(..., mode="edge")` is a thing, we could use it instead of `np.pad` in a couple of places:

```
$ git grep -n mode=\'edge\'
scipy/fft/tests/test_real_transforms.py:40:    y2 = xp.asarray(np.pad(np.asarray(y), pad, mode='edge'))
scipy/fft/tests/test_real_transforms.py:111:    y2 = xp.asarray(np.pad(np.asarray(y), pad, mode='edge'))
scipy/ndimage/_interpolation.py:219:            padded = np.pad(input, npad, mode='edge')
scipy/signal/_short_time_fft.py:1096:            'edge': dict(mode='edge'),
```
 
Re `dct` et al: Ok thanks. There are CuPy/jax delegation targets (https://docs.cupy.dev/en/stable/reference/generated/cupyx.scipy.fft.dct.html and https://docs.jax.dev/en/latest/_autosummary/jax.scipy.fft.dct.html), so it's a potential enhancement. Not worth a dedicated issue though.
